ml4h/evaluate_segmentations.py

```diff
@@ -0,0 +1,141 @@
+        self.offset = 0
+        self.limit = len(files)
+    #
+    def __len__(self):
+        return self.limit
+    #
+    def __getitem__(self, index):
+        self.offset += 1
+        x, y, _, b = next(self.generator)
+        return x, y, [None]
+    #
+    def on_epoch_end(self):
+        self.offset = 0
+
+
+# Given a TensorFlow model, a dictionary of metrics, and a data generator
+# or (x,y) pair, we evaluate the model and return a pair of dictionary for
+# per-batch (n=1) metrics and overall mean metrics.
+# 
+# As a preprocessing step the input `model` must already be re/compiled
+# to use the exact metrics in the `metrics` list.
+def evaluate_collect_metrics(model: tf.keras.models.Model, 
+        data_generator: list, 
+        metrics: List[Callable] = None):
+    # Reformat metrics list to a metrics dictionary of the form
+    # {name: function pointer}.
+    if metrics is not None:
+        if isinstance(metrics, list):
+            metrics_dict = {m.name: m for m in metrics}
+        else:
+            metrics_dict = metrics
+    else:
+        metrics_dict = {m.name: m for m in model.metrics}
+    # Reload model
+    model.compile(optimizer=model.optimizer,
+                        loss=model.loss,
+                        metrics=list(metrics_dict.values()))
+    #
+    # Logger for outputs
+    logger = BatchMetricsLogger(metrics = metrics_dict)
+    try:
+        eval = model.evaluate(data_generator, callbacks=[logger])
+    except Exception as e:
+        raise Exception(f'Failed to evaluate model:\n{e}')
+    eval_batch = pd.DataFrame(logger.storage, index = np.arange(len(logger.storage)))
+    # Workaround for old Keras
+    eval = {out: eval[i] for i, out in enumerate(model.metrics_names)}
+    eval = pd.DataFrame(eval,index=[0])
+    # Make predictions
+    prediction_prob = model.predict(data_generator)
+    # Assumes channel_order
+    predictions = tf.argmax(prediction_prob, -1)
+    # Return eval, eval_batch pair
+    return eval, eval_batch, prediction_prob, predictions
+
+
+def evaluate_segmentation_models(models: List[tf.keras.models.Model], 
+        data_generators: list, 
+        metrics: List[Callable] = None,
+        channel_order: str = 'channels_last'): # Todo: Currently unused
+    # `metrics` must be either a list of callable metrics functions
+    if len(models) != len(data_generators):
+        if len(data_generators) != 1:
+            raise ValueError("Unequal number of models and data generators. Must be equal in numbers or use only a single data generator for all models.")
+        if len(data_generators) == 1:
+            data_generators = data_generators * len(models)
+    #
+    # Reformat metrics list to a metrics dictionary of the form
+    # {name: function pointer}.
+    if metrics is not None:
+        metrics_dict = {m.name: m for m in metrics}
+    #
+    model_count = 0
+    evaluations = list()
+    for m,g in zip(models, data_generators):
+        if metrics is None:
```

Comment:
What is the difference between what happens in line 107 and what happens here

Reply:
It should be such that 1) if `metrics` is passed then used that for all models, otherwise 2) foreach model use whatever metrics it came with (`model.metrics`)